internal/core/metadata/operators/device_profile/value_descriptor.go

```diff
@@ -0,0 +1,72 @@
+package device_profile
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+// ValueDescriptorAdder provides the necessary functionality for creating a ValueDescriptor.
+type ValueDescriptorAdder interface {
+	Add(vdr *contract.ValueDescriptor, ctx context.Context) (string, error)
+}
+
+// ValueDescriptorAddExecutor creates ValueDescriptor(s) via the operator pattern.
+type ValueDescriptorAddExecutor interface {
+	Execute() error
+}
+
+type addValueDescriptor struct {
+	ctx    context.Context
+	drs    []contract.DeviceResource
+	client ValueDescriptorAdder
+	logger logger.LoggingClient
+}
+
+// Execute creates the necessary ValueDescriptors for a set of DeviceResources.
+func (a addValueDescriptor) Execute() error {
+	for _, dr := range a.drs {
+
+		desc := createValueDescriptorFromDeviceResource(dr)
+		id, err := a.client.Add(&desc, a.ctx)
+		if err != nil {
```

Comment:
You might consider integrating the [ErrServiceClient](https://github.com/edgexfoundry/go-mod-core-contracts/blob/5f45a867aa5be00d1f023143816e99a17fae2b1b/clients/types/errors.go#L36) here so that the error returned from core-data is bubbled out to the caller adding the DeviceProfile.

Reply:
Won't the `Add` method return an `ErrServiceClient`? If my assumption is wrong then what should we set the `StatusCode` and/or `body` information?

Reply:
That's just one of several different error types you might expect from this call. Drilling down to [PostRequest](https://github.com/edgexfoundry/go-mod-core-contracts/blob/5f45a867aa5be00d1f023143816e99a17fae2b1b/clients/request.go#L103) I see two different error types that could be returned, and higher up the stack there are others.

Looking at this again, this area is fine. It's probably in `restAddDeviceProfile` where we need type checking in order to return the appropriate http status. The following `err != nil` check
```
op := device_profile.NewAddExecutor(r.Context(), vdc, LoggingClient, dp.DeviceResources...)
err = op.Execute()
if err != nil {
	http.Error(w, err.Error(), http.StatusInternalServerError)
	return
}
```
For now I'd be satisfied if this just did a type check w/r/t `ErrServiceClient`. If the error is of that type, then return the status code and message wrapped by the error. Otherwise return a 500 as it does now.